mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,513 @@
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&self.0.to_string())
+    }
+}
+
+impl<'de> Deserialize<'de> for PeerAddress {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct PeerAddressVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for PeerAddressVisitor {
+            type Value = PeerAddress;
+
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str("a string representation of peer id.")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<PeerAddress, E>
+            where
+                E: serde::de::Error,
+            {
+                if value.len() > 100 {
+                    return Err(serde::de::Error::invalid_length(
+                        value.len(),
+                        &"peer id cannot exceed 100 characters.",
+                    ));
+                }
+
+                Ok(mm2_libp2p::PeerId::from_str(value)
+                    .map_err(serde::de::Error::custom)?
+                    .into())
+            }
+
+            fn visit_string<E>(self, value: String) -> Result<PeerAddress, E>
+            where
+                E: serde::de::Error,
+            {
+                self.visit_str(&value)
+            }
+        }
+
+        deserializer.deserialize_str(PeerAddressVisitor)
+    }
+}
+
+#[derive(Debug, Display)]
+enum SignValidationError {
+    #[display(
+        fmt = "Healthcheck message is expired. Current time in UTC: {now_secs}, healthcheck `expires_at` in UTC: {expires_at_secs}"
+    )]
+    Expired { now_secs: u64, expires_at_secs: u64 },
+    #[display(
+        fmt = "Healthcheck message have too high expiration time. Max allowed expiration seconds: {max_allowed_expiration_secs}, received message expiration seconds: {remaining_expiration_secs}"
+    )]
+    LifetimeOverflow {
+        max_allowed_expiration_secs: u64,
+        remaining_expiration_secs: u64,
+    },
+    #[display(fmt = "Public key is not valid.")]
+    InvalidPublicKey,
+    #[display(fmt = "Signature integrity doesn't match with the public key.")]
+    FakeSignature,
+    #[display(fmt = "Process failed unexpectedly due to this reason: {reason}")]
+    Internal { reason: String },
+}
+
+impl HealthcheckMessage {
+    pub(crate) fn generate_message(ctx: &MmArc, is_a_reply: bool) -> Result<Self, String> {
+        let p2p_ctx = P2PContext::fetch_from_mm_arc(ctx);
+        let keypair = p2p_ctx.keypair();
+        let sender_public_key = keypair.public().encode_protobuf();
+
+        let data = HealthcheckData {
+            sender_public_key,
+            expires_at_secs: u64::try_from(Utc::now().timestamp()).map_err(|e| e.to_string())?
+                + healthcheck_message_exp_secs(),
+            is_a_reply,
+        };
+
+        let signature = try_s!(keypair.sign(&try_s!(data.encode())));
```

Comment:
It's a bit strange that we encode the `sender_public_key` using `protobuf` but we use `messagepack` for the whole data. It's fine as it will cause no problems but thought to bring it up. If you wanna encode the whole data using protobuf you can follow the swaps v2 examples, it's up to you @onur-ozkan
https://github.com/KomodoPlatform/komodo-defi-framework/blob/7723b50fcb533ebe12b621eb862b8a958113696b/mm2src/mm2_main/src/lp_swap.rs#L1706

Reply:
IMO protobuf definitions add unnecessary complexity since we don't need to provide a language-agnostic interface.

Reply:
Protobuf has some backward compatibility benefits if some structs are nested as opposed to messagepack ref. https://github.com/KomodoPlatform/komodo-defi-framework/pull/1017#discussion_r675524636 , if this struct won't change in this way then no need for it for sure. Please resolve this if it's not needed.

Reply:
We could create versioning enumeration like

```rs
enum SomeMessage {
Version1 {...},
Version2 {...},
Version3 {...},
...
}
```

Considering a node with 3 versions in this struct and an old node with only 2 versions, can't old node decode Version1/Version2 messages with messagepack ?

p.s.: We don't need such support for healthchecks; I just wonder how messagepack can't handle versioned messages.

Reply:
> Considering a node with 3 versions in this struct and an old node with only 2 versions, can't old node decode Version1/Version2 messages with messagepack ?

I did some experiments with message pack: apparently ser/deser old enum to/from extended enum works okay if the order of variants preserved in the extended enum (if new items mixed between old items then ser/deser old to/from extended fails)

Reply:
This approach can be used:

```rs
#[repr(u8)]
enum Demo {
    V1 { x: i32 } = 1,
    V2 { x: i32 } = 3,
    V3 { x: i32 } = 2
}
```

to make sure ordering never changes.

Reply:
I think what important is ordering in the enum definition (messagepack encode enums as maps with key = variant index).
Like for these enums ser/deser should work (ofc apart from trying to deser Extended::Ver3 to Old):
```
#[derive(Clone, Debug, Eq, Deserialize, PartialEq, Serialize)]
enum Old {
    Ver1(u32),
    Ver2(String),
}
#[derive(Clone, Debug, Eq, Deserialize, PartialEq, Serialize)]
enum Extended {
    Ver1(u32),
    Ver2(String),
    Ver3{ x: u32 },
}
```

And this should fail:
```
#[derive(Clone, Debug, Eq, Deserialize, PartialEq, Serialize)]
enum Old {
    Ver1(u32),
    Ver2(String),
}
#[derive(Clone, Debug, Eq, Deserialize, PartialEq, Serialize)]
enum Extended {
    Ver3{ x: u32 },
    Ver1(u32),
    Ver2(String),
}
```